docs/proposals/rebalancing-clusters-across-shards-dynamically.md

```diff
@@ -0,0 +1,144 @@
+* It will check if any application controller is flagged as Unhealthy or has not updated its status in Redis during the heartbeat process for a certain period of time.
+* If the status for an application controller was already flagged as Unhealthy, it will pick one of the clusters from its list and assign the cluster to itself.
+* If the status is not flagged and an application controller has not updated the last active timestamp in a long time, then we mark the Application Controller as Unhealthy. 
+
+This will continue to happen till the time the list of clusters is marked as empty. As soon as the list is marked as empty, the entry of the shard/application-controller is removed from Redis.
+
+*Note:* We will continue to use the cluster to shard assignment approach being used today.
+
+### Pros
+* Every Application Controller would be able to take action on finding the distribution of load.
+* Every Application Controller will monitor the status of Unhealthy shard and would be able to take action or flag for action.
+
+### Cons
+
+* Possibility of race conditions while flagging the shard as Unhealthy during the heartbeat process. Although this can be handled using the [distributed locks](https://redis.io/docs/manual/patterns/distributed-locks/) in Redis.
+
+* In scenarios when Redis becomes unavailable, the heartbeat mechanism will pause working till the redis comes back online again. This will also pause the dynamic redistribution of clusters till Redis comes back online. The redistribution of clusters will be triggered again when Redis comes back online.
+
+
+### Security Considerations
+
+* This would be a breaking change of converting StatefulSets to Deployments. Any automation done by customers which is based on the assumption that the controller is modelled as a StatefulSet would break with this change. 
+
+* We would rely on Redis to store the current Application Controller to Shard mapping. In case the Redis is not available, it would not affect the regular working of ArgoCD. The dynamic distribution of clusters among healthy shards would stop working with the heartbeat process till Redis comes back up online, but the application controllers will continue managing their workloads.
+
+
+### Upgrade / Downgrade Strategy
+
+* Working ArgoCD itself should not affected. An initial restart of all the application controller pods is expected when we switch from StatefulSet to Deployment or vice-versa.
+
+* There would be some initial delays in the reconciliation process during the transistion from StatefulSet to Deployment. If someone is not using sharding at all, they should not face any issues.
+
+## Alternatives
+
+An alternative approach would be to use Leader Election strategy. By implementing leader election, one of the healthy application controllers will be appointed as leader. The leader controller will be responsible for assigning clusters to the shards and balancing load across the shards.
```

Comment:
I much prefer the leader election strategy as I think it will avoid dealing with complicated edge cases and is a simpler implementation depending less on a specific Redis feature.

Reply:
I cannot imagine how a leader can determine if the controller is healthy or not. It will need to watch pods, and evaluate the health of each pod which sounds complicated. 

"Claiming" shards in each pod independently looks more reliable to me.

Reply:
We can get past the Redis feature by using ConfigMap as suggested by Alex in another comment. I also think that having each pod claim a shard would be beneficial and avoid drawbacks of leader election, especially when leader becomes unhealthy or starts performing distribution in a wrong manner.